src/ipc_listener.cpp

```diff
@@ -208,9 +260,13 @@ int zmq::ipc_listener_t::set_address (const char *addr_)
 
 int zmq::ipc_listener_t::close ()
 {
-    zmq_assert (s != retired_fd);
-    int rc = ::close (s);
-    errno_assert (rc == 0);
+    int rc;
+
+    if ( s != -1 ) {
+      zmq_assert (s != retired_fd);
```

Comment:
On Mon, Mar 14, 2016 at 10:51 AM, Luca Boccassi
<notifications@github.com> wrote:
> Yes, and that's why, I think, the code as-is did a return -1 instead of goto err if open_socket failed.
>
> We need the assert to enforce that this API is not used improperly, IE: called on a socket that's already closed. Do you think it's possible to leave that in with your changes?

I may need to rework it a bit.  What I was trying to do was cleanup
the mkdtemp() on an error.  I thought that re-using
ipc_listener_t::close() was the best way to do that.

I can revert ipc_listener_t::close(), and update
ipc_listener_t::set_address() to properly cleanup the directory if
open_socket() fails.  I understand your API checks, and I'll fix
things on my end.

Are the prohibitions on the use of exceptions?  I initially coded it
to throw an exception in create_wildcard_address(), but didn't see
handling of exceptions elsewhere.  If exceptions are ok, I can rework
set_address() to make sure that there are no stray files or
directories upon failure pretty easily.  Otherwise, the code may look
a bit ugly.

Pete